### ISSUE TITLE
EF 7 not loading my entities when using Eager or Explicit loading (Blazor Server)...

### ISSUE THREAD

**Issue by sipi41** [dotnet/efcore#30943]
Hello and thank you for reading... I think I found a bug and I would really appreciate any help if I'm missing something, it would be nice if you point out what the problem is.

I do have a class called **FACT_Entry**, which has a property List<**FactComment**> comments, each of these comments has a User (AppIdentityUser : IdentityUser), and they are linked through a key of type string. to simplify my classes are something like this: (reduced for simplicity)

```c#
public class FACT_Entry
{
    public Guid tool_guid { get; set; }
    public **virtual List<FactComment> comments** { get; set; } = new();
    public string tool_name { get; set; }
}
public class FactComment
    {
        public Guid comment_guid { get; set; }
        public **Guid tool_guid** { get; set; }       
        public string Comment { get; set; }
        public **string? created_by** { get; set; }
        public **virtual AppIdentityUser User** { get; set; }
    }
public class AppIdentityUser : IdentityUser
    {
        public virtual List<FactComment> FactComments { get; set; }
    }
```

In other words... each FACT tool has a list of comments, and each comment has the ID of the Identity user that created it, represented by AppIdentityUser class that inherits IdentityUser. I'm trying to load a single tool, with its comments and the user who created it, so I created something like this:

```c#
            using TSMC_ODS_DbContext db = await odsContextFactory.CreateDbContextAsync(); //create my db context from factory

            Editable_FACT = await db.FACT
                                .Include(FACT => FACT.comments)
                                .ThenInclude(comment => comment.User)                                
                                .FirstOrDefaultAsync(f => f.tool_name == "KEMGH5"); //KEMGH5 just a random name
```

**THE PROBLEM** is that I get the list of comments, but nothing from the ThenInclude, (no User, is null...) I did check if the SQL generated from EF and it returned the list as follows:
![image](https://github.com/dotnet/efcore/assets/688562/2a32a660-a2a1-4675-a37b-c24a2e9f926b)

I did check my relationships and all are fine, indeed if not, EF could not find the info I want but it does, as we saw on the image.

I tried to use explicit loading but the result is the same, is this a bug? am I missing something? any help on this will be appreciated, I spent hours already not sure what to do here... As usual, **I cannot send you my whole project and database**, as this belongs to my boss, but I can offer you remote access if you want to come and play by to find out what's going on... **thank you for any help you may bring.**

**Comment by roji** [comment]
@sipi41 the best way to create a minimal repro is to start a new console application project from scratch, and to try to do the minimum in order to trigger the bug. You typically take only the part of your EF context which you need, and tweak the code so as to make it fail in the same way as in your real project. Very frequently, as part of this process users discover a bug in their own code, and the issue is resolved.

Doing this is highly recommended (and frequently required) before submitting a bug report, otherwise the maintainers have no way of investigating your problem.

**Comment by sipi41** [comment]
Thank you @roji I see... I'll copy my project, not sure how to do the db thing... how do you upload the db ?? ... I'm thinking, this is nice for a very small project like mine.. what if this is a project that has advanced too much? (many files, many parts, etc)... ? it could be impossible... there should be a better way than this... just my opinion.

**Comment by roji** [comment]
@sipi41 we're very interested in investigating all bug reports that we get - this is one reason why we have an open-source github repo, and if you'll look at the various issues that get filed you'll see us responding and interacting with all of them.

However, we get dozens of bug reports per day; if the engineers on the team did a call with each and every person opening an issue, we'd basically no longer have any time to do actually build features.

In modern software engineering, you're expected to be able to isolate the bug that you're seeing, and submit that as a minimal repro. You don't upload your DB; you create a minimal DB schema - just the table(s) needed, with just the data needed. You do the same with code: a simple console program, with just the parts needed to trigger the bug. This process of isolating or extracting the repro allows users to better understand how your code and EF interact, and very frequently leads users to find their bugs on their own. And if there's an actual bug in EF, this leads to a repro which we can then use to investigate and fix the issue.

> because sometimes its just time-consuming do a "sample project"

Keep in mind that if it's too time-consuming for you to do a sample project, you're basically asking the EF team to do the work; except that you're the one actually seeing the bug, and so in a much better position to actually do that.

To summarize, most users reporting issues on our repo manage do provide sample projects (or repros) - that's just the most efficient way to work on bugs, both for them and for us. Of course, you can pay for Microsoft support, at which point you can talk with a Microsoft support engineer who will look over your code with you.

**Comment by sipi41** [comment]
ok @roji I'll create a sample project, and make a folder with the db file inside to be attached to whatever server they want to use, which will be a reduced copy of the db so you can restore, it will take me a couple of hours, hope that's enough.

**Comment by roji** [comment]
Sure. The simpler you make it, the better. For the database, you can even just create it with EF itself (e.g. EnsureCreated) and insert the required data, or just include a quick SQL script that does CREATE TABLE and inserts the required data. That's much better than attaching a database backup or similar.

**Comment by sipi41** [comment]
Dear friends @roji and @ajcvickers, I have created a public repository (if this should not be public please let me know, as I'm not an expert in repositories), please access it with the following route:  **https://github.com/sipi41/EF_NOT_LOADING_RELATED_DATA**

Hope to hear from you soon, thank you for all the help you may bring! and please, please let me know if something does not go as expected or if you have problems, I cleaned the code the most I could as this has thousands of parts...

**Comment by sipi41** [comment]
Thank you guys, @ajcvickers and @roji ! that was the problem, Identity without asking saved the keys as strings, guid like but as strings... you mentioned .net 8 could provide this automatically but did not find any docs... would be ok to migrate to .net 8? how do I make sure these things are working as expected?  Thank you so much! you are the heroes of the day!

**Comment by ajcvickers** [comment]
>  you mentioned .net 8 could provide this automatically but did not find any docs

https://github.com/dotnet/efcore/issues/27526
